File: app/src/main/java/com/github/goldy1992/mp3player/service/library/content/retriever/SongsFromFolderRetriever.java
package com.github.goldy1992.mp3player.service.library.content.retriever;

import android.content.ContentResolver;
import android.database.Cursor;
import android.os.Handler;
import android.provider.MediaStore;

import com.github.goldy1992.mp3player.commons.MediaItemType;
import com.github.goldy1992.mp3player.service.library.content.filter.SongsFromFolderResultsFilter;
import com.github.goldy1992.mp3player.service.library.content.parser.SongResultsParser;

import javax.inject.Inject;
import javax.inject.Named;

import static com.github.goldy1992.mp3player.service.library.content.Projections.SONG_PROJECTION;

public class SongsFromFolderRetriever extends ContentResolverRetriever {

    private static final String WHERE = MediaStore.Audio.Media.DATA + " LIKE ? ";

    @Inject
    public SongsFromFolderRetriever(ContentResolver contentResolver,
                                    SongResultsParser resultsParser,
                                    @Named("worker") Handler handler,
                                    SongsFromFolderResultsFilter songsFromFolderResultsFilter) {
        super(contentResolver, resultsParser, handler, songsFromFolderResultsFilter);
    }

    @Override
    public MediaItemType getType() {
        return MediaItemType.SONG;
    }

    @Override
    Cursor performGetChildrenQuery(String id) {
        String[] whereArgs = {id + "%"};
        return contentResolver.query(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI ,getProjection(),
                WHERE, whereArgs, null);
    }

    @Override
    String[] getProjection() {
        return SONG_PROJECTION.toArray(new String[0]);
    }

}

File: service/src/main/java/com/github/goldy1992/mp3player/service/library/content/retriever/SongsFromFolderRetriever.kt
package com.github.goldy1992.mp3player.service.library.content.retriever

import android.content.ContentResolver
import android.database.Cursor
import android.os.Handler
import android.provider.MediaStore
import android.provider.MediaStore.MediaColumns
import com.github.goldy1992.mp3player.commons.MediaItemType
import com.github.goldy1992.mp3player.service.library.content.Projections.SONG_PROJECTION
import com.github.goldy1992.mp3player.service.library.content.filter.SongsFromFolderResultsFilter
import com.github.goldy1992.mp3player.service.library.content.parser.SongResultsParser
import javax.inject.Inject
import javax.inject.Named

class SongsFromFolderRetriever @Inject constructor(contentResolver: ContentResolver,
                                                   resultsParser: SongResultsParser,
                                                   @Named("worker") handler: Handler,
                                                   songsFromFolderResultsFilter: SongsFromFolderResultsFilter?) : ContentResolverRetriever(contentResolver, resultsParser, handler, songsFromFolderResultsFilter) {
    override val type: MediaItemType?
        get() = MediaItemType.SONG

    public override fun performGetChildrenQuery(id: String?): Cursor {
        val whereArgs = arrayOf("$id%")
        return contentResolver.query(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI, projection,
                WHERE, whereArgs, null)
    }

    override val projection: Array<String?>?
        get() = SONG_PROJECTION.toTypedArray()

    companion object {
        private const val WHERE = MediaColumns.DATA + " LIKE ? "
    }
}